### PR TITLE
Make MachineEnv a per-ABI property

### PR DESCRIPTION
**Pull request by uweigand**
The MachineEnv structure contains the allocatable and preferred register sets.  This is currently fixed per TargetIsa - however, conceptually these register sets can differ between ABIs on the same ISA.

To allow for this, replace the TargetIsa machine_env routine with an ABIMachineSpec get_machine_env routine.  To ensure the structure is still only allocated once, cache it via static OnceLock variables.

No functional change intended.

FYI @cfallin - this is what I suggested in last week's meeting.

### CODE REVIEW

**Review by cfallin — APPROVED** [review]
LGTM, thanks a bunch!

(stealing review from @elliottt since discussed in previous Cranelift meeting, hope that's ok!)

**Comment by cfallin** [comment]
@uweigand happy to merge once the merge conflict is resolved...

**Comment by uweigand** [comment]
> @uweigand happy to merge once the merge conflict is resolved...

I've rebased now.

**Pull request merged**
